### PR TITLE
cephadm: handle exceptions applying extra services during bootstrap

### PR DESCRIPTION
**Pull request by adk3798**
Otherwise we risk hitting a mismatch between the cephadm binary version and the container image version we're bootstrapping on, resulting in bootstrap failing. Example in the tracker.

Fixes: https://tracker.ceph.com/issues/59082

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/cephadm/cephadm.py [hunk 1]
```diff
@@ -5422,12 +5422,20 @@ def prepare_ssh(
     if not ctx.skip_monitoring_stack:
         for t in ['ceph-exporter', 'prometheus', 'grafana', 'node-exporter', 'alertmanager']:
             logger.info('Deploying %s service with default placement...' % t)
-            cli(['orch', 'apply', t])
+            try:
+                cli(['orch', 'apply', t])
+            except Exception:
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:5427]: Can we be more specific with the exception being caught? Catching all `Exception`s is a wide net and could mask other issues.

Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:5427]: I just checked the sources. call_throws will raise RuntimeError if call returns a non-success return code, so catching RuntimeError alone should be sufficient.

File: src/cephadm/cephadm.py [hunk 2]
```diff
@@ -5422,12 +5422,20 @@ def prepare_ssh(
...
 
     if ctx.with_centralized_logging:
         for t in ['loki', 'promtail']:
             logger.info('Deploying %s service with default placement...' % t)
-            cli(['orch', 'apply', t])
+            try:
+                cli(['orch', 'apply', t])
+            except Exception:
+                logger.error(f'Failed to apply service type {t}. '
+                             'Perhaps the ceph version being bootstrapped does not support it')
```
Comment (rkachach) [inline src/cephadm/cephadm.py:5438]: should we exit with an error user code in this case as one of the options provided by the user are failing? .. capturing the exception will lead to a "good" bootstrap when not all the provided options are not fulfilled.

File: src/cephadm/cephadm.py (hunk 1 above)
Comment (adk3798) [reply src/cephadm/cephadm.py]: swapped to only catching `RuntimeError`

File: src/cephadm/cephadm.py (hunk 2 above)
Comment (adk3798) [reply src/cephadm/cephadm.py]: changed it to now set `ctx.error_code` when one of these failures happens, which is the same mechanism being used currently for returning a nonzero error code when applying a provided spec fails.

**Review by phlogistonjohn — APPROVED** [review]
LGTM, thanks!

**Review by rkachach — APPROVED** [review]
LGTM

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-03-15_01:34:51-orch:cephadm-wip-adk-testing-2023-03-14-1905-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-03-30_12:19:06-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/

After reruns, 6 failures:

- 2 failed staggered upgrade test, combination of https://tracker.ceph.com/issues/58535 plus the attempted fix not working
- 2 failed nfs-keepalive-only tests, newly added by a PR in the run, and not an issue in terms of merging any other PR
- 1 instance of https://tracker.ceph.com/issues/49287 (known issue)
- 1 failure in test_nfs task. The failure only happens in the test variation with the agent on (https://pulpito.ceph.com/adking-2023-03-29_11:57:55-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/7225304 passed) so I believe this is a timing thing related to the agent. Nothing to block merging on anything not specifically related to nfs.

**Comment by adk3798** [comment]
jenkins test make check

**Pull request merged**